encodings/fastlanes/src/bitpacking/compute/take.rs

```diff
@@ -4,16 +4,31 @@ use fastlanes::BitPacking;
 use itertools::Itertools;
 use vortex::array::{PrimitiveArray, SparseArray};
 use vortex::compute::{slice, take, TakeFn};
-use vortex::{Array, ArrayDType, IntoArray, IntoArrayVariant};
+use vortex::{Array, ArrayDType, IntoArray, IntoArrayVariant, IntoCanonical};
 use vortex_dtype::{
     match_each_integer_ptype, match_each_unsigned_integer_ptype, NativePType, PType,
 };
 use vortex_error::VortexResult;
 
 use crate::{unpack_single_primitive, BitPackedArray};
 
+// assuming the buffer is already allocated (which will happen at most once) then unpacking
+// all 1024 elements takes ~8.8x as long as unpacking a single element on an M2 Macbook Air.
+// see https://github.com/spiraldb/vortex/pull/190#issue-2223752833
+const UNPACK_CHUNK_THRESHOLD: usize = 8;
+const PATCH_THRESHOLD: usize = 64;
+
 impl TakeFn for BitPackedArray {
     fn take(&self, indices: &Array) -> VortexResult<Array> {
+        // If the indices are large enough, it's faster to flatten and take the primitive array.
+        if indices.len() * UNPACK_CHUNK_THRESHOLD > self.len() {
```

Comment:
I don't think the UNPACK_CHUNK_THRESHOLD is necessarily the right factor. Suppose there are 8K indices but they're all <= 1024. These would get chunked together and only trigger decompression of the first chunk.

When the indices are "chunk-contiguous" (i.e. all indices in [0, 1024) come first, then [1024, 2048), etc.), `take_primitive` already makes the correct choice between decompressing a chunk and individuals.

However, regardless of order, if the indices are sufficiently large, manipulating them dominates runtime. There's basically a per-index cost to construct `relative_indices` and another per-index cost to copy the patches to the right place. It's something like

```
if (n_indices * (chunk_and_relativize_cost + patch_cost * sparsity))
   >
   (decompress_cost * n_values)
```

Reply:
Yeah, I agree that this is a bit crude / could be further refined, but it's a decent approximation.

The reason that I picked this same threshold thing is that if the indices are distributed uniformly then this will be ~optimal (even if they're sorted; but definitely if they're not).

The worst case would be if they're all in [0, 1024)... will add a bench for that